pkg/queryfrontend/shard_query.go

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+// This is a modified copy from
+// https://github.com/cortexproject/cortex/blob/master/pkg/querier/queryrange/split_by_interval.go.
+
+package queryfrontend
+
+import (
+	"context"
+
+	"github.com/cortexproject/cortex/pkg/querier/queryrange"
+	"github.com/thanos-io/thanos/pkg/querysharding"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+)
+
+// PromQLShardingMiddleware creates a new Middleware that shards PromQL aggregations using grouping labels.
+func PromQLShardingMiddleware(queryAnalyzer *querysharding.QueryAnalyzer, numShards int, limits queryrange.Limits, merger queryrange.Merger) queryrange.Middleware {
+	return queryrange.MiddlewareFunc(func(next queryrange.Handler) queryrange.Handler {
+		return querySharder{
+			next:          next,
+			limits:        limits,
+			queryAnalyzer: queryAnalyzer,
+			numShards:     numShards,
+			merger:        merger,
+		}
+	})
+}
+
+type querySharder struct {
+	next   queryrange.Handler
+	limits queryrange.Limits
+
+	queryAnalyzer *querysharding.QueryAnalyzer
+	numShards     int
+	merger        queryrange.Merger
+}
+
+func (s querySharder) Do(ctx context.Context, r queryrange.Request) (queryrange.Response, error) {
+	analysis, err := s.queryAnalyzer.Analyze(r.GetQuery())
+	if err != nil {
+		return nil, err
+	}
+
+	if !analysis.IsShardable() {
+		return s.next.Do(ctx, r)
+	}
+
+	reqs := s.shardQuery(r, analysis)
+
+	reqResps, err := queryrange.DoRequests(ctx, s.next, reqs, s.limits)
```

Comment:
Why the sharding feature is implemented on the query frontend? It is also possible to do the same thing on the querier side directly. WDYT?

Reply:
The reason for this is because we need to split the query and unify the result, and this is already possible with query frontend. If we added it to the querier, then that querier has to talk to other queriers which is additional complexity that we need to introduce.

Reply:
👍🏽

Reply:
Can be added later.